### ISSUE TITLE
Could we charge per-pixel like PICO-8 does to "scope" games?

### ISSUE THREAD

**Issue by joshgoebel** [nesbox/TIC-80#1788]
Creating a new issue for this to separate it from the "specific hardware is slow" discussion.  

Let's forget about the language runtime internals themselves (PICO-8 counts Lua opcode expense, for example), but should we add cost accounting to the TIC API functions themselves and then set a "CPU" cap?  If so, I might be interested in taking this up after the WASM stuff if finished.  Have a look at how PICO-8 does this: https://pico-8.fandom.com/wiki/CPU

I think if we just did this for drawing functions alone we might end up in a world where cartridges performed consistently across many different hardware specs - at the cost of not allowing people to do crazy things on ultra-fast hardware (like they can now).  And of course there would be a discussion of how these limits should work (or not) on older cartridges that weren't designed with the limits.

This makes the large assumption that most cartridge are slow for drawing reasons rather than spending too much time doing math or crazy loops, etc.  That might be unfounded for some games.

Thoughts?

---

The original post in the other thread:

I think perhaps this is the fault of our games (and that we have no limits and don't set any expectations of what people should be doing)...  I got curious and just dug into PICO-8 cost accounting: https://pico-8.fandom.com/wiki/CPU

For example you have "139,810 cycles per frame at 60 FPS"... and sprites cost per-pixel * 2... so an 8x8 sprites is 64 pixels, so 128 cycles.... so if look at bunnymark we bunny are 4x rendered, making each bunny 32x32... giving it a PICO-8 cycle cost of 2,048. If we just plug this into PICO-8 we'd find we could have 68 bunnies per frame at 100% CPU burn. (that's assuming zero cost for all the Lua code itself)

>  and 75 sprites is a pretty good amount

So I think based on PICO-8 at least that this statement isn't too far off.  I can draw 700-800 bunnies on my laptop (Debug build) at 60 fps.

> From maybe 10 games that I tried only Witch 'em Up and Cauliflower Power runs smooth.

I wonder if you couldn't provide a bit more detail about which exact games - and could anyone else on the platform confirm the same slow results?  What about playing `panda`, is it also slow?  I guess I'm just wondering WHY they are so slow... and saying that 75 sprites per frame isn't "slow"... are they slow because they are using 200 sprites, or for some other reason?

_Originally posted by @joshgoebel in https://github.com/nesbox/TIC-80/issues/1739#issuecomment-1007614861_

**Comment by joshgoebel** [comment]
Having thought about this a little more I think for drawing heavy carts all we need to do is add a cost to each drawing function.  We could just make it the approx. number of pixels drawn (since that's so trivial to calculate).  So all the drawing functions get "expensed"... when you run out of budget you've used "too much CPU".  We could of course represent this as a little graph (green/orange/red) or some such.  _This doesn't have to be any kind of hard limit_, it could just be used to let someone know their cartridge may not perform well on all systems, etc.

Truly CPU bound/math carts are harder - and would require per-language work I think... but I'd wager (someone correct me if I'm wrong) that most carts are draw operation bound...

So I'd say I just need a good sample of:

- lightweight game
- medium weight game
- SUPER intensive heavy weight game

Then based on that we can get some actual benchmarks and decide what a "reasonable" threshold should be.  I don't think this type of cost accounting should be that hard to add at all.

**Comment by nesbox** [comment]
The idea reminds me "draw calls" on modern GPUs and the most difficult question I see here is to calculate how much each drawing function costs.

Btw, I think FPS80 is a good example of a heavy weight game https://tic80.com/play?cart=237 :)

**Comment by joshgoebel** [comment]
>  and the most difficult question

True, but assuming we're not pushing the hardware close to it's limits we don't have to be exact here, just reasonable-ish.  For example I think something like this would be a reasonable start:

- `pix` costs 1
- `rect` costs width * height
- `spr` costs 8x8, plus 30% penalty for reading sprite data
- `textri` costs pixels written plus 60% penalty for reading texture, tracking border pixels, etc (and how complex it is)
- `cls` costs 240*136 (but with say a 60% discount for full screen clears that use memset)

I'm just guessing at the penalties but you could also quickly get really good ballparks with measurement:

- use `drawRect` to draw the full screen (x times), get a metric
- use `spr` to draw the screen full of sprites (x times), get a metric

However much slower `spr` is than `rect` is it's penalty (and now it's very close to actually measuring ~ CPU use).  Of course all the routines that use pixel border buffers would have penalties, etc.

I think you'd go a long ways without penalties, but if you wanted to be fair to someone writing a sprite bound game vs a first person shooter then you need to spend a little time figuring out the relative speed of operations, not just the raw pixel count... a ray caster might only draw 240*136 pixels (well plus some magin for overlap) but how expensive our line drawing routines are vs just writing pixels sequentially is going to matter a lot more.

So for starters I think it'd be easier if we exposed "expense" via the API... this would give us a chance to collect some stats, let people build their perf tools (inside the games), etc... before making it "official"... some type of `cpuusage()` or `stat()` call?

**Comment by joshgoebel** [comment]
OK, `fps80` is a great example of what seems to be an actual CPU/runtime bound game (vs drawcalls).  I'm not sure such a game is really a candidate for playable on "all hardware" (at least not without huge optimizations).  It's using 12-16ms per frame on my 2.5GHz laptop.  Ie, it's using 75-100% CPU of my whole system just for the render engine.  This isn't a case of someone overtaxing the sprite drawing routines a bit... are most ambitious games on the platform people writing custom 3D rendering engines in Lua?  

_If so I don't think simple performance metrics (like I imagined) are going to help that much._   Or perhaps it would force/incentivize them to write more optimized code??  I'm guessing it could be optimized since I know there are such games on PICO-8 that manage to fit within their own tight imaginary CPU constraints.

_What I'm proposing wouldn't be able to accurately measure this game._  For that we'd need a `ticker` inside Lua... same overall idea though... a "reasonable" number of overall ticks/opcodes... vs how many ticks per frame it actually cost.  PICO-8 does per opcode accounting, but I think that may be above our need, a more complex patch against Lua.  But honestly if you do a measurement with `time()` and you're using more than 25% of a modern CPU I'd first work on that metric before getting any more nuanced.

**Comment by joshgoebel** [comment]
Ah it seems we already have tic_mem `impl.studio.tic` (`tic80_local`) that's used for drawing studio stuff, with the system palette coming from `&getConfig()->cart->bank0`... so perhaps we should just reuse this?

**Comment by joshgoebel** [comment]
For example I start in `studioInit()`

```
impl.tic80local = (tic80_local*)tic80_create(impl.samplerate);
impl.studio.tic = impl.tic80local->memory;
```

Then very quickly that calls:

```
    initConfig(impl.config, impl.studio.tic, impl.fs);
    initStart(impl.start, impl.studio.tic, args.cart);
```

Passing them `tic_mem*` which is fine, because we have that.  But then both initConfig AND initStart call:

```
tic_api_reset(tic)
```

Which again is fine because it takes  `tic_mem*`... but immediately inside tic_api_reset we cast tic_mem to tic_core and start zeroing its `state`, etc...  but there is no state... `tic80_local` doesn't include state... what am I missing or is this truly a latent bug?

**Comment by joshgoebel** [comment]
Ah, finally in `tic80_create`...

```
tic80->memory = tic_core_create(samplerate);
```

`tic_core_create` wrongly IMHO returning a `tic_mem*`....

This is all very confusing for someone trying to read and understand the code... I mean the "trick" of typecasting pointers back and forth is one thing - perhaps it has utility, but shouldn't data structs always refer to the LARGER struct, ie if we have a struct:

```
struct {
  a* widget
} thing;
```

Where `a` is just a tiny thing... but then also we **commonly** do:

```
wholeapp* b = (wholeapp*) thing.a;
```

It's it much clearer to write:

```
struct {
  a* wholeapp;
} thing;
```

Now immediately that `a` is a pointer NOT just to a widget, but the whole application.  Knowing that widget is the first pointer allows me to actually backtrack it, find out that it's struct might start with a `widget`, etc... but it's much harder to do the opposite.

**Comment by nesbox** [comment]
> `tic80_local` doesn't have any state struct... what am I missing

It's another level of abstraction, in addition to the **core** and the **studio** we also have a thing called `player` which can only play carts with a very simple API `tic80_create() tic80_load() tic80_tick() tic80_delete()`, and `tic80_local` is designed to hide all the internal kitchen, pls look at the `src/system/sdl/player.c` for example.
Honestly, we don't need to use `tic80_local` in the studio at all and we could replace 
``` c
impl.tic80local = (tic80_local*)tic80_create(impl.samplerate);
impl.studio.tic = impl.tic80local->memory;
```
with
```c
impl.studio.tic = tic_core_create(samplerate);
// and remove tic80local from the studio
```
to remove confusion
